### ISSUE TITLE
about dataloader

### ISSUE THREAD

**Issue by Breezewrf** [xuebinqin/U-2-Net#167]
hello xuebinqin：
     thanks to your fanscinated project, I am curious about your tricks on image-preprocess, it seems there is a great performance improvement after the dataloader, but I don't quite understand the idea of that part of the code, could you please give me a general idea?
     thanks a lot

**Comment by vincentzhang** [comment]
The preprocessing was mostly defined in the transform argument to the SalObjDataset. It performs rescaling followed by random cropping. 
Another part of the preprocessing was offline augmentation by horizontal flipping, which is described in the paper.
Hope it helps.

**Comment by Breezewrf** [comment]
> The preprocessing was mostly defined in the transform argument to the SalObjDataset. It performs rescaling followed by random cropping.
> Another part of the preprocessing was offline augmentation by horizontal flipping, which is described in the paper.
> Hope it helps.

Thank you very much for your reply. I'm going to use cpp to reproduce the predicting process. After reading your python code, it seems that the main preprocessing in the predict part should be normal normalization? However, it is strange that the effect will be a little worse after the same cpp implementation.(Maybe there are some details omitted)

**Comment by vincentzhang** [comment]
Thanks for providing more details.
Sorry that my previous answer wasn't complete. The pre-processing also includes normalization followed by standardization.
But it doesn't seem to me that this part of pre-processing would explain any difference between cpp and python implementation since it's arithmetic operators plus max/min. 
Without access to the cpp code, I cannot tell but two sources that might be worth looking into:
1. the resizing function, whether it's similar to the skimage implementation
2. the neural network implementation. It would not be too surprising if pytorch cpp and python API are slightly different. But I have no experience in it.

**Comment by Breezewrf** [comment]
> Thanks for providing more details.
> Sorry that my previous answer wasn't complete. The pre-processing also includes normalization followed by standardization.
> But it doesn't seem to me that this part of pre-processing would explain any difference between cpp and python implementation since it's arithmetic operators plus max/min.
> Without access to the cpp code, I cannot tell but two sources that might be worth looking into:
> 
> 1. the resizing function, whether it's similar to the skimage implementation
> 2. the neural network implementation. It would not be too surprising if pytorch cpp and python API are slightly different. But I have no experience in it.

Thank you very much for your guidance!. I have modified two details after read and debug again and again,finally,fortunately the accuracy has improved a lot. One is the reset of the input image. 320 * 320 is more suitable than the 224 * 224 I used before. The disadvantage is that the speed has decreased. On the other hand, it is bgr2rgb. It seems that the effect of the model will be improved after converting to RGB in advance. After making up for the above two details, the output result is almost the same as that of the python version, but only the edge is not as smooth as that of the python version, which may be due to the difference in accuracy.

Thank you again
